sw/host/opentitanlib/src/util/openocd.rs

```diff
@@ -0,0 +1,63 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+
+use std::fs::File;
+use std::io::Write;
+use std::process::{Command, Stdio, Child};
+use crate::io::jtag::{Jtag, JtagParams};
+
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::impl_serializable_error;
+
+/// Represents an OpenOCD server and we can interact with interactively
+pub struct OpenOcdClient {
+    /// OpenOCD child process
+    child: Child,
+}
+
+/// Errors related to the OpenOCD client.
+#[derive(Error, Debug, Deserialize, Serialize)]
+pub enum OpenOcdClientError {
+    #[error("No OpenOCD binary path provided")]
+    OpenOcdBinaryRequired,
+    #[error("Generic error {0}")]
+    Generic(String),
+}
+impl_serializable_error!(OpenOcdClientError);
+
+impl OpenOcdClient {
+    pub fn new(opts: &JtagParams) -> Result<OpenOcdClient> {
+        log::info!("Starting OpenOCD (binary: {})...", &opts.openocd);
+        let mut cmd = Command::new(&opts.openocd);
+        if let Some(cfg) = &opts.openocd_adapter_config {
+            cmd.args(["-f", &cfg]);
+        }
+        let child = cmd.stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn().context("failed to spawn openocd")?;
+        Ok(OpenOcdClient {
+            child
+        })
+    }
+
+    fn connect_telnet(&self) -> Result<()> {
+        Ok(())
+    }
+}
+
+impl Drop for OpenOcdClient {
+    fn drop(&mut self) {
+        self.child.kill();
```

Comment:
Is this an unhandled `Result`?

```suggestion
        self.child.kill().ok();
```

It would be nice to send OpenOCD a `SIGTERM` (or a close command over the telnet?) and `.wait()` for it to exit. `SIGKILL` might leave a corrupted connection to the FPGA, for example.

Reply:
This is just unfinished code. The proper way will be to first send "shutdown" to openocd, and kill it if it does not terminate after a time-out. This will be implemented once I have the telnet code.